File: src/shim/ftop_signal.c
#include <errno.h>
#include <signal.h>
#include <sys/types.h>
#include <unistd.h>

enum {
  FTOP_SIGNAL_WINCH = 1,
  FTOP_SIGNAL_INT = 2,
  FTOP_SIGNAL_TERM = 3,
  FTOP_SIGNAL_TSTP = 4,
  FTOP_SIGNAL_CONT = 5,
  FTOP_SIGNAL_HUP = 6,
  FTOP_SIGNAL_KILL = 7,
  FTOP_SIGNAL_STOP = 8,
  FTOP_SIGNAL_USR1 = 9,
  FTOP_SIGNAL_USR2 = 10
};

static volatile sig_atomic_t ftop_winch_pending = 0;
static volatile sig_atomic_t ftop_int_pending = 0;
static volatile sig_atomic_t ftop_term_pending = 0;
static volatile sig_atomic_t ftop_tstp_pending = 0;
static volatile sig_atomic_t ftop_cont_pending = 0;

static void ftop_set_error(int *sys_errno, int code) {
  if (sys_errno != NULL) *sys_errno = code;
}

static int ftop_signal_number_impl(int signal_id) {
  switch (signal_id) {
  case FTOP_SIGNAL_WINCH:
    return SIGWINCH;
  case FTOP_SIGNAL_INT:
    return SIGINT;
  case FTOP_SIGNAL_TERM:
    return SIGTERM;
  case FTOP_SIGNAL_TSTP:
    return SIGTSTP;
  case FTOP_SIGNAL_CONT:
    return SIGCONT;
  case FTOP_SIGNAL_HUP:
    return SIGHUP;
  case FTOP_SIGNAL_KILL:
    return SIGKILL;
  case FTOP_SIGNAL_STOP:
    return SIGSTOP;
  case FTOP_SIGNAL_USR1:
    return SIGUSR1;
  case FTOP_SIGNAL_USR2:
    return SIGUSR2;
  default:
    return 0;
  }
}

static void ftop_signal_handler(int signo) {
  switch (signo) {
  case SIGWINCH:
    ftop_winch_pending = 1;
    break;
  case SIGINT:
    ftop_int_pending = 1;
    break;
  case SIGTERM:
    ftop_term_pending = 1;
    break;
  case SIGTSTP:
    ftop_tstp_pending = 1;
    break;
  case SIGCONT:
    ftop_cont_pending = 1;
    break;
  default:
    break;
  }
}

static int ftop_install_signal(int signo) {
  struct sigaction action;

  action.sa_handler = ftop_signal_handler;
  sigemptyset(&action.sa_mask);
  action.sa_flags = 0;
  return sigaction(signo, &action, NULL);
}

int ftop_signal_setup(void) {
  if (ftop_install_signal(SIGWINCH) != 0) return -1;
  if (ftop_install_signal(SIGINT) != 0) return -1;
  if (ftop_install_signal(SIGTERM) != 0) return -1;
  if (ftop_install_signal(SIGTSTP) != 0) return -1;
  if (ftop_install_signal(SIGCONT) != 0) return -1;
  return 0;
}

int ftop_signal_check(int signal_id) {
  switch (signal_id) {
  case FTOP_SIGNAL_WINCH:
    return ftop_winch_pending != 0;
  case FTOP_SIGNAL_INT:
    return ftop_int_pending != 0;
  case FTOP_SIGNAL_TERM:
    return ftop_term_pending != 0;
  case FTOP_SIGNAL_TSTP:
    return ftop_tstp_pending != 0;
  case FTOP_SIGNAL_CONT:
    return ftop_cont_pending != 0;
  default:
    return 0;
  }
}

void ftop_signal_clear(int signal_id) {
  switch (signal_id) {
  case FTOP_SIGNAL_WINCH:
    ftop_winch_pending = 0;
    break;
  case FTOP_SIGNAL_INT:
    ftop_int_pending = 0;
    break;
  case FTOP_SIGNAL_TERM:
    ftop_term_pending = 0;
    break;
  case FTOP_SIGNAL_TSTP:
    ftop_tstp_pending = 0;
    break;
  case FTOP_SIGNAL_CONT:
    ftop_cont_pending = 0;
    break;
  default:
    break;
  }
}

int ftop_signal_suspend_self(void) {
  struct sigaction action;

  action.sa_handler = SIG_DFL;
  sigemptyset(&action.sa_mask);
  action.sa_flags = 0;
  if (sigaction(SIGTSTP, &action, NULL) != 0) return -1;
  if (raise(SIGTSTP) != 0) return -1;
  return ftop_install_signal(SIGTSTP);
}

int ftop_signal_number(int signal_id) {
  return ftop_signal_number_impl(signal_id);
}

int ftop_current_pid(void) {
  return (int)getpid();
}

int ftop_kill(int pid, int signal_number, int *sys_errno) {
  ftop_set_error(sys_errno, 0);
  if (pid <= 0 || signal_number <= 0) {
    ftop_set_error(sys_errno, EINVAL);
    return -1;
  }

  if (kill((pid_t)pid, signal_number) != 0) {
    ftop_set_error(sys_errno, errno);
    return -1;
  }

  return 0;
}

File: src/shim/ftop_signal.f90
module ftop_signal
  use, intrinsic :: iso_c_binding, only : c_int
  implicit none
  private

  integer, parameter, public :: FTOP_SIGNAL_WINCH = 1
  integer, parameter, public :: FTOP_SIGNAL_INT = 2
  integer, parameter, public :: FTOP_SIGNAL_TERM = 3
  integer, parameter, public :: FTOP_SIGNAL_TSTP = 4
  integer, parameter, public :: FTOP_SIGNAL_CONT = 5
  integer, parameter, public :: FTOP_SIGNAL_HUP = 6
  integer, parameter, public :: FTOP_SIGNAL_KILL = 7
  integer, parameter, public :: FTOP_SIGNAL_STOP = 8
  integer, parameter, public :: FTOP_SIGNAL_USR1 = 9
  integer, parameter, public :: FTOP_SIGNAL_USR2 = 10

  public :: ftop_current_pid
  public :: ftop_kill
  public :: ftop_signal_check
  public :: ftop_signal_clear
  public :: ftop_signal_number
  public :: ftop_signal_setup
  public :: ftop_signal_suspend_self

  interface
    integer(c_int) function c_ftop_signal_setup() bind(C, name="ftop_signal_setup")
      import :: c_int
    end function c_ftop_signal_setup

    integer(c_int) function c_ftop_signal_check(signal_id) bind(C, name="ftop_signal_check")
      import :: c_int
      integer(c_int), value :: signal_id
    end function c_ftop_signal_check

    subroutine c_ftop_signal_clear(signal_id) bind(C, name="ftop_signal_clear")
      import :: c_int
      integer(c_int), value :: signal_id
    end subroutine c_ftop_signal_clear

    integer(c_int) function c_ftop_signal_suspend_self() bind(C, name="ftop_signal_suspend_self")
      import :: c_int
    end function c_ftop_signal_suspend_self

    integer(c_int) function c_ftop_signal_number(signal_id) bind(C, name="ftop_signal_number")
      import :: c_int
      integer(c_int), value :: signal_id
    end function c_ftop_signal_number

    integer(c_int) function c_ftop_current_pid() bind(C, name="ftop_current_pid")
      import :: c_int
    end function c_ftop_current_pid

    integer(c_int) function c_ftop_kill(pid, signal_number, sys_errno) bind(C, name="ftop_kill")
      import :: c_int
      integer(c_int), value :: pid
      integer(c_int), value :: signal_number
      integer(c_int), intent(out) :: sys_errno
    end function c_ftop_kill
  end interface

contains

  logical function ftop_signal_setup() result(success)
    success = c_ftop_signal_setup() == 0_c_int
  end function ftop_signal_setup

  logical function ftop_signal_check(signal_id) result(pending)
    integer, intent(in) :: signal_id

    pending = c_ftop_signal_check(int(signal_id, c_int)) /= 0_c_int
  end function ftop_signal_check

  subroutine ftop_signal_clear(signal_id)
    integer, intent(in) :: signal_id

    call c_ftop_signal_clear(int(signal_id, c_int))
  end subroutine ftop_signal_clear

  logical function ftop_signal_suspend_self() result(success)
    success = c_ftop_signal_suspend_self() == 0_c_int
  end function ftop_signal_suspend_self

  integer function ftop_signal_number(signal_id) result(signal_number)
    integer, intent(in) :: signal_id

    signal_number = int(c_ftop_signal_number(int(signal_id, c_int)))
  end function ftop_signal_number

  integer function ftop_current_pid() result(pid)
    pid = int(c_ftop_current_pid())
  end function ftop_current_pid

  logical function ftop_kill(pid, signal_number, error_code) result(success)
    integer, intent(in) :: pid
    integer, intent(in) :: signal_number
    integer, intent(out), optional :: error_code
    integer(c_int) :: sys_errno

    success = c_ftop_kill(int(pid, c_int), int(signal_number, c_int), sys_errno) == 0_c_int
    if (present(error_code)) error_code = int(sys_errno)
  end function ftop_kill

end module ftop_signal
